Router executor prepare support for non-partition columns

metdos opened the pull request:
Implements router executor part of #572.

Adds regression tests for #645.

metdos commented:
One note about composite types:

Currently, for composite types, we send text oid because on the worker node composite type oid will be different. Due to this trick, we get a casting error after the 5th execution.

Example prepare statement:
```sql
PREPARE prepared_composite(test_composite_type) AS
INSERT INTO composite_type_partitioned_table VALUES  (1, $1);

EXECUTE prepared_composite('(4, 9)');
```

Error on the worker nodes:
```
LOG:  execute <unnamed>: INSERT INTO composite_type_partitioned_table_102153 (id, composite_column) VALUES (1, '(4,9)'::test_composite_type)
DETAIL:  parameters: $1 = '(4,9)'
ERROR:  column "composite_column" is of type test_composite_type but expression is of type text at character 87
HINT:  You will need to rewrite or cast the expression.
```

Fortunately, there is a workaround. Simply, we can provide fields of the composite type instead of the compact version above. Such as;
```sql
PREPARE prepared_composite_workaround(integer, integer) AS
INSERT INTO composite_type_partitioned_table VALUES  (1, ROW($1, $2));

EXECUTE prepared_composite_workaround(4, 9);
```

metdos commented:
Hey @anarazel, could you take a high-level look? Then, @sumedhpathak will do the low-level code review.

Thanks!

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -475,12 +479,58 @@ ReturnRowsFromTuplestore(uint64 tupleCount, TupleDesc tupleDescriptor,
-	querySent = PQsendQuery(connection, query);
+	if (paramListInfo != NULL)
+	{
+		int parameterIndex = 0;
```
Can we put the contents of this in a separate function? It seems likely that we'll need it in other places too, and it's quite reasonably standalone.

anarazel replied:
To clarify, I'm only talking about the parts that build parameterTypes/Values, not the PQsendQueryParams itself.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -475,12 +479,58 @@ ReturnRowsFromTuplestore(uint64 tupleCount, TupleDesc tupleDescriptor,
+			 * Use text oid for data types where the oid values can be different
+			 * on the master and worker nodes.
+			 */
+			if (parameterData->ptype >= FirstNormalObjectId)
```
Might be worthwhile to generalize the function from cab03a6274b77087ec5ec9b7d4e13a5506b7f422 instead.

metdos replied:
> Might be worthwhile to generalize the function from cab03a6 instead.

This is a bit different than the COPY version. Here, for any of user-defined types, we can't use the original oid of the type because it will be different on master and worker nodes. For example for hll type, COPY version returns false and continues using the text format, but here we need to change the oid.

anarazel commented on src/backend/distributed/planner/multi_router_planner.c:
```diff
@@ -745,7 +745,17 @@ ExtractInsertPartitionValue(Query *query, Var *partitionColumn)
+		if (IsA(targetEntry->expr, Param))
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot execute a prepared insert statement on "
```
I think that's a pretty confusing message, and it's also not necessarily true. I wonder if we shouldn't just always error out if there's a param on the partition key instead. That'll be a lot easier for users to understand.

metdos replied:
> I wonder if we shouldn't just always error out if there's a param on the partition key instead. That'll be a lot easier for users to understand.

Mostly, users' drivers are running PREPARE, and generally, these drivers run the prepared statement just once. Also, we have unnamed prepared statements workaround. For reference #306. Erroring at the first trial will limit our coverage.

Given this, I would prefer to error out after 5th iteration. I will think about the error message, and I'm open to suggestions. What do you think?

metdos replied:
Actually, while rebasing, I noticed that the issue #645 is fixed now in the planner with new commits of @lithp.
https://github.com/citusdata/citus/blame/master/src/backend/distributed/planner/multi_router_planner.c#L308

And the current error message is:
ERROR:  values given for the partition column must be constants or constant expressions

I will keep the new regression tests but will remove the fix.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -475,12 +479,58 @@ ReturnRowsFromTuplestore(uint64 tupleCount, TupleDesc tupleDescriptor,
+			 */
+			if (parameterData->ptype >= FirstNormalObjectId)
+			{
+				parameterTypes[parameterIndex] = TEXTOID;
```
Hm. I think this should be 0 instead, which means the type should be automatically inferred. There's not necessarily a cast from text to the other side of the operation, so using text isn't a good idea here.

metdos replied:
> Hm. I think this should be 0 instead, which means the type should be automatically inferred. There's not necessarily a cast from text to the other side of the operation, so using text isn't a good idea here.

Using 0 was the my initial attempt looking into the [documentation](https://www.postgresql.org/docs/9.5/static/libpq-exec.html) of PQexecParams(). However, if I use 0, this errors out at the even first trial on the worker node.

```
ERROR:  could not determine data type of parameter $1
STATEMENT:  INSERT INTO helloworld_102225 (id, set) VALUES (1, '\x118b7f'::hll)
```

There are a couple of places where this error message is used in PostgreSQL source. I will check which one is fired, and try to find a solution.

metdos replied:
This is the code part which errors out:
https://github.com/postgres/postgres/blob/master/src/backend/tcop/postgres.c#L1345

Most likely, parse_analyze_varparams() just above this check fails to infer the data type. Should we add some casting functions to new data type (hll here) to make it work?
https://github.com/postgres/postgres/blob/master/src/backend/tcop/postgres.c#L1330

anarazel replied:
The reason this fails is that the parameter is unused in your example, because standard_planner() inlined its use. The fix is simple: Murat's router executor change deparses the query using the original Query structure, instead of the modified one. Which means $1 etc. are still used, and thus the types can be inferred.

metdos commented:
@anarazel, I think that I have addressed your comments. The last commit changes the regression tests, and regression tests fail currently. After changes of @mtuncer are rebased, regression tests will pass.

One note about composite types: Currently, if I create a SELECT prepared statement on a composite type and provide text for the composite type, it fails.

```sql

PREPARE prepared_composite_select(test_composite_type) AS
SELECT count(*) FROM composite_type_partitioned_table WHERE id = 1 AND composite_column = $1::test_composite_type;

EXECUTE prepared_composite_select('(4, 9)');
```

```
ERROR:  input of anonymous composite types is not implemented
STATEMENT:  SELECT count(*) AS count FROM composite_type_partitioned_table_102230 WHERE ((id = 1) AND (composite_column = $1))
```

For the prepared statement, we don't pass the composite type casting to the worker nodes for SELECT queries. The problem would be on the router executor part. I didn't check details on the behalf of other tasks. Inserts and unparameterized composite types work fine. For a workaround on SELECTs, we can use individual parts of the composite type. Such as;

```sql

PREPARE prepared_composite_select(int, int) AS
SELECT count(*) FROM composite_type_partitioned_table WHERE id = 1 AND composite_column = ROW($1, $2);

EXECUTE prepared_composite_select(4, 9);
```

Also note that, other custom types such as hll work fine.

ozgune commented:
@metdos -- Since we're tracking the v5.2 milestone as part of issues (#572 / #645), I'm removing it from this pull request. If this pull request addresses an unopen issue, please add back the milestone tag.

anarazel commented:
@metdos The composite type param issue can easily enough be worked around by the user adding a ```$1::composite_type``` style. So I think that's ok for now.  It's imo a postgres issue that that's required.

metdos commented:
> @metdos The composite type param issue can easily enough be worked around by the user adding a $1::composite_type style.

That's the problem. Even though the user uses $1::composite_type style, we drop casting part while we are sending the query to the worker nodes.

I'm copying the example master query and the worker query.

```sql

PREPARE prepared_composite_select(test_composite_type) AS
    SELECT count(*) FROM composite_type_partitioned_table WHERE id = 1 AND 
        composite_column = $1::test_composite_type;

EXECUTE prepared_composite_select('(4, 9)');
```

```
ERROR:  input of anonymous composite types is not implemented
STATEMENT:  SELECT count(*) AS count FROM composite_type_partitioned_table_102230 
     WHERE ((id = 1) AND (composite_column = $1))
```

Maybe I can open a different issue for this? What do you think?

metdos commented:
@anarazel, @sumedhpathak 

I rebased the code after changes of Murat on the master node. Could you take a look?

anarazel commented:
@metdos That's weird - not very long ago, when I manually merged, the cast worked quite well.

anarazel commented:
Anyway, I think the change looks ok. We should indeed open a separate issue for the composite type issue.

metdos commented:
@anarazel, squashing and merging. I will also open an issue about the composite type casting.
